owncloudApp/src/main/java/com/owncloud/android/ui/activity/PassCodeActivity.kt

```diff
@@ -0,0 +1,425 @@
+    }
+
+    public override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        outState.putBoolean(KEY_CONFIRMING_PASSCODE, mConfirmingPassCode)
+        outState.putStringArray(KEY_PASSCODE_DIGITS, mPassCodeDigits)
+    }
+
+    private inner class PassCodeDigitTextWatcher(private val mIndex: Int, private val mLastOne: Boolean) : TextWatcher {
+        private operator fun next(): Int {
+            return if (mLastOne) 0 else mIndex + 1
+        }
+
+        /**
+         * Performs several actions when the user types a digit in an input field:
+         * - saves the input digit to the state of the activity; this will allow retyping the
+         * pass code to confirm it.
+         * - moves the focus automatically to the next field
+         * - for the last field, triggers the processing of the full pass code
+         *
+         * @param s     Changed text
+         */
+        override fun afterTextChanged(s: Editable) {
+            if (s.length > 0) {
+                if (!mConfirmingPassCode) {
+                    mPassCodeDigits[mIndex] = mPassCodeEditTexts[mIndex]!!.text.toString()
+                }
+                mPassCodeEditTexts[next()]!!.requestFocus()
+                if (mLastOne) {
+                    processFullPassCode()
+                }
+            } else {
+                Timber.d("Text box $mIndex was cleaned")
+            }
+        }
+
+        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
+            // nothing to do
+        }
+
+        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
+            // nothing to do
+        }
+
+        /**
+         * Constructor
+         *
+         * @param index         Position in the pass code of the input field that will be bound to
+         * this watcher.
+         * @param lastOne       'True' means that watcher corresponds to the last position of the
+         * pass code.
+         */
+        init {
+            require(mIndex >= 0) {
+                "Invalid index in " + PassCodeDigitTextWatcher::class.java.simpleName +
+                        " constructor"
+            }
+        }
+    }
+
+    companion object {
+        const val ACTION_REQUEST_WITH_RESULT = "ACTION_REQUEST_WITH_RESULT"
+        const val ACTION_CHECK_WITH_RESULT = "ACTION_CHECK_WITH_RESULT"
+        const val ACTION_CHECK = "ACTION_CHECK"
+        const val KEY_PASSCODE = "KEY_PASSCODE"
+        const val KEY_CHECK_RESULT = "KEY_CHECK_RESULT"
+
+        // NOTE: PREFERENCE_SET_PASSCODE must have the same value as settings_security.xml-->android:key for passcode preference
+        const val PREFERENCE_SET_PASSCODE = "set_pincode"
+        const val PREFERENCE_PASSCODE = "PrefPinCode"
+
+        // NOTE: This is required to read the legacy pin code format
+        const val PREFERENCE_PASSCODE_D = "PrefPinCode"
+
+        const val numberOfPassInputs = 4
```

Comment:
A constant does not fix completely the problem. It could be the first iteration, we can generate an APK with any number requirement but it does not fix the problem at all. If we want to make it brandable, we need to perform more changes, so I think this is not the best approach.

From my point of view, it should get the number from the `setup.xml`, by default `4`. So if they introduce a new field in ownbrander to select the length of the passcode, we will be ready and no more changes will be required from our side 👍

Reply:
You are right, yet we are in a hurry to get this out to the customer. I'd propose to leave it like this and make it brandable later down the line. I'll create a new issue for this.

Reply:
Ok, to be done in https://github.com/owncloud/android/issues/3358